data_source.go

```diff
@@ -99,40 +120,90 @@ func Start(ds DataSource) error {
 		return err
 	}
 
+	ds.RunDoneActivate()
 	if err := ds.StartRun(); err != nil {
 		return err
 	}
 
-	ds.RunDoneAdd()
-	// Have the DataSource produce data until graceful stop.
-	go func() {
-		defer ds.RunDoneDone()
-		for {
-			// fmt.Println("calling blockingRead")
-			if err := ds.blockingRead(); err == io.EOF {
-				log.Println("blockingRead returns io.EOF, more stopping the source")
-				// EOF error should occur after abortSelf has been closed
-				// ds.CloseOutputs() // why is this here, ds.Stop also calls CloseOutputs
+	go CoreLoop(ds, queuedRequests)
```

Comment:
Would this be more clear as

```
go func() {
defer ds.RunDoneDeative()
CoreLoop(ds,queuedRequests)
}()
```

Reply:
No. It would be less clear. I'll comment the `RunDoneActivate()` call instead.